diff --git a/train_local_obst.sh b/train_local_obst.sh
index fe3c221..6cbda9d 100644
--- a/train_local_obst.sh
+++ b/train_local_obst.sh
@@ -1,6 +1,6 @@
 python -m swarm_rl.train \
 --env=quadrotor_multi --train_for_env_steps=1000000000 --algo=APPO --use_rnn=False \
---num_workers=4 --num_envs_per_worker=4 --learning_rate=0.0001 --ppo_clip_value=5.0 --recurrence=1 \
+--num_workers=200 --num_envs_per_worker=8 --learning_rate=0.0001 --ppo_clip_value=5.0 --recurrence=1 \
 --nonlinearity=tanh --actor_critic_share_weights=False --policy_initialization=xavier_uniform \
 --adaptive_stddev=False --with_vtrace=False --max_policy_lag=100000000 --rnn_size=256 \
 --gae_lambda=1.00 --max_grad_norm=5.0 --exploration_loss_coeff=0.0 --rollout=128 --batch_size=1024 \
@@ -14,4 +14,4 @@ python -m swarm_rl.train \
 --quads_neighbor_encoder_type=no_encoder --quads_neighbor_visible_num=2 \
 --quads_use_obstacles=True --quads_obst_spawn_area 8 8 --quads_obst_density=0.2 --quads_obst_size=0.6 \
 --quads_obst_collision_reward=5.0 --quads_obstacle_obs_type=octomap --quads_use_downwash=True \
---experiment=test_obst_avoid
\ No newline at end of file
+--experiment=test_obst_avoid
